apollo-federation/src/sources/connect/json_selection/selection_set.rs

```diff
@@ -0,0 +1,421 @@
+//! Functions for applying a [`SelectionSet`] to a [`JSONSelection`]. This creates a new
+//! `JSONSelection` mapping to the fields on the selection set, and excluding parts of the
+//! original `JSONSelection` that are not needed by the selection set.
+
+use std::collections::HashSet;
+
+use apollo_compiler::executable::Selection;
+use apollo_compiler::executable::SelectionSet;
+
+use crate::sources::connect::json_selection::Alias;
+use crate::sources::connect::json_selection::NamedSelection;
+use crate::sources::connect::JSONSelection;
+use crate::sources::connect::Key;
+use crate::sources::connect::PathSelection;
+use crate::sources::connect::SubSelection;
+
+/// Find all fields in a selection set with a given name.
+fn find_fields_in_selection_set<'a>(set: &'a SelectionSet, name: &str) -> Vec<&'a Selection> {
+    let mut fields = vec![];
+    for selection in &set.selections {
+        match selection {
+            Selection::Field(field) => {
+                if field.name == name {
+                    fields.push(selection);
+                }
+            }
+            Selection::FragmentSpread(_) => {
+                // Ignore - these are always expanded into inline fragments
+            }
+            Selection::InlineFragment(fragment) => {
+                fields.extend(find_fields_in_selection_set(&fragment.selection_set, name));
+            }
+        }
+    }
+    fields
+}
+
+impl JSONSelection {
+    pub fn apply_selection_set(&self, selection_set: &SelectionSet) -> Self {
+        match self {
+            Self::Named(sub) => Self::Named(sub.apply_selection_set(selection_set)),
+            Self::Path(path) => Self::Path(path.apply_selection_set(selection_set)),
+        }
+    }
+}
+
+impl SubSelection {
+    fn apply_selection_set(&self, selection_set: &SelectionSet) -> Self {
+        let mut new_selections = Vec::new();
+        let mut dropped_fields = HashSet::new();
+        let mut referenced_fields = HashSet::new();
+        for selection in &self.selections {
+            match selection {
+                NamedSelection::Field(alias, name, sub) => {
+                    let key = alias
+                        .as_ref()
+                        .map(|a| a.name.to_string())
+                        .unwrap_or(name.to_string());
+                    let fields = find_fields_in_selection_set(selection_set, &key);
+                    if self.star.is_some() {
+                        if fields.is_empty() {
+                            dropped_fields.insert(key.clone());
+                        } else {
+                            referenced_fields.insert(key.clone());
+                        }
+                    }
+                    for field in fields {
+                        let field_response_key = field
+                            .as_field()
+                            .map(|s| s.response_key())
+                            .unwrap()
```

Comment:
We've mostly avoided `.unwrap()` in this part of the codebase (except in tests), so you might want to hoist the `if let Some(field) = field.as_field() {...}`  to the top of the loop here (and possibly double-check your usage of `.unwrap()` here and elsewhere).

Reply:
Done, and added `deny` for this to prevent future changes from adding any `unwrap` or `expect`.